[optimizer] introduce optimizer functional API, refactor Adagrad

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45221 [dist_optim] introduce distributed functional optimizer
* #44791 [optimizer] refactor Adam to use functional API
* **#44715 [optimizer] introduce optimizer functional API, refactor Adagrad**

We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

Differential Revision: [D23935258](https://our.internmc.facebook.com/intern/diff/D23935258)

wanchaol committed:
[WIP][optimizer] introduce functional API for optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][optimizer] introduce functional API for optimizer"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

vincentqb commented on torch/optim/adagrad.py:
```diff
@@ -63,40 +64,29 @@ def step(self, closure=None):
-                    p.addcdiv_(grad, std, value=-clr)
+            for p in group['params']:
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
```
could these new lists become memory expensive?

wanchaol replied:
I think it's fine, those lists are just holding intrusive_ptr copies of the tensors, which are pretty small.

vincentqb commented on torch/optim/functional.py:
```diff
@@ -0,0 +1,46 @@
+from torch import Tensor
+from typing import List
+
+# TODO: use foreach API in optim.functional to do all the computation
```
note to self: we'll need to make a conscious effort to avoid having too many versions of the same optimizers, with the risk of having them out of sync.

cc  #43507

vincentqb replied:
offline meeting with @wanchaol and @izdeby: #43507 will be merged first and then updated to functional form later.

vincentqb commented on torch/optim/functional.py:
```diff
@@ -0,0 +1,46 @@
+
+# TODO: use foreach API in optim.functional to do all the computation
+
+def adagrad(params: List[Tensor],
```
we should have a test that gives some guarantees we have not diverged from the original implementation.

wanchaol replied:
currently I am relying on the original optimizer tests to test the coverage. https://github.com/pytorch/pytorch/blob/24eea364f761bc841cb3f3890d1ce0200571856e/test/test_optim.py#L384 Are there any new cases that you think we need to add to give the guarantees?

vincentqb replied:
Since we are replacing the current implementation, the C++ tests confirm that the the python and C++ give the same answers, see [comment](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986). This gives the signal we need here.

cc @anjali411

vincentqb commented on torch/optim/functional.py:
```diff
@@ -0,0 +1,46 @@
+
+        clr = lr / (1 + (step - 1) * lr_decay)
+
+        if grad.is_sparse:
```
note to self: I have seen users suggesting splitting out sparse/dense versions, see [comment](https://github.com/pytorch/pytorch/pull/30232#issuecomment-671203290). My current impression is that, given the scope of the current pull request, I would indeed prefer keeping both as done here instead of pushing the refactoring further..

vincentqb commented:
also relates to #39279

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

vincentqb commented:
One more observation: from this PR, the [C++ implementation](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986) doesn't get a functional implementation. Is that something you planned to do @wanchaol ?

cc @anjali411

vincentqb commented:
Oh right, we should add that to the documentation :)

wanchaol commented:
> One more observation: from this PR, the [C++ implementation](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986) doesn't get a functional implementation. Is that something you planned to do @wanchaol ?
> 

@vincentqb yes I can make follow up PRs to make C++ APIs to also include functional APIs


> Oh right, we should add that to the documentation :)

do you mean that we should add the functional APIs to the documentation like this? https://pytorch.org/docs/stable/nn.functional.html

vincentqb commented:
> do you mean that we should add the functional APIs to the documentation like this? https://pytorch.org/docs/stable/nn.functional.html

We could also simply add them [here](https://pytorch.org/docs/stable/optim.html), though the page is getting long :)

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

vincentqb reviewed (approved):
I second @albanD, the PR LGTM. We can do C++, doc, etc in a separate one later.

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] introduce optimizer functional API, refactor Adagrad"


We have provided a nice and intuitive API in Python. But in the context of large scale distributed training (e.g. Distributed Model Parallel), users often want to use multithreaded training instead of multiprocess training as it provides better resource utilization and efficiency.

This PR introduces functional optimizer concept (that is similar to the concept of `nn.functional`), we split optimizer into two parts: 1. optimizer state management 2. optimizer computation. We expose the computation part as a separate functional API that is available to be used by internal and OSS developers, the caller of the functional API will maintain their own states in order to directly calls the functional API. While maintaining the end user API be the same, the functional API is TorchScript friendly, and could be used by the distributed optimizer to speed up the training without GIL.

[ghstack-poisoned]

vincentqb commented:
The documentation is in #45513.

vincentqb commented:
> I second @albanD, the PR LGTM. We can do C++, doc, etc in a separate one later.

@wanchaol, can you clarify if there were other steps needed that you had in mind so that we can move this beyond [prototype](https://github.com/pytorch/pytorch/pull/45513/files#diff-d5dc35b4fad1f8158d7ea192701b8241d6e08aa77ba7ce31fae97b767799d582R298)?

wanchaol commented:
> @wanchaol, can you clarify if there were other steps needed that you had in mind so that we can move this beyond [prototype](https://github.com/pytorch/pytorch/pull/45513/files#diff-d5dc35b4fad1f8158d7ea192701b8241d6e08aa77ba7ce31fae97b767799d582R298)?

@vincentqb I would say improving the coverage of functional optimizers, to cover most of commonly used optimizers (which I am planning to do before 1.8 release), and improve the test coverage, this will make it be becomes beta to use.

To mark this as stable, we need to cover all existing optimizers that could be turned into functional optimizer, and possibly integrate with the `foreach` api.
